pxtcompiler/emitter/service.ts

```diff
@@ -1207,6 +1207,33 @@ namespace ts.pxtc.service {
                 return Object.keys(tooltipOrBlock).map(k => (<pxt.Map<string>>tooltipOrBlock)[k]).join(" ");
             };
 
+            // Fill default parameters in block string
+            const computeBlockString = (symbol: SymbolInfo): string => {
+                let block = symbol.attributes.block;
+                if (symbol.attributes?.paramDefl) {
+                    const paramDefl = symbol.attributes.paramDefl;
+                    for (let param in paramDefl) {
+                        if (block.indexOf(`%${param}`) >= 0) {
```

Comment:
Don't use the block string directly, use the parsed block definition instead. It's stored in _def and _expandedDef on the attributes object.

Also %, is the old way of defining block definitions. All future blocks should use $

Reply:
Any parameters using % are not guaranteed to match the name in paramDefl. That's part of the reason why we switched to the new syntax. In the old way you could put whatever name you wanted into the block string (it was purely based off index)

Reply:
Actually, you might have more luck using https://github.com/microsoft/pxt/blob/master/pxtlib/blocks.ts#L105 which resolves all the param names correctly

Reply:
hmm i don't know if the full compile info is worth it, right now if i don't find the matching name i'm just appending the default to the end of the block string and the fuzzy search is generous enough that it mostly works. although i guess it would be good to keep all our "constructing string from blockdef" code in one place